### PR TITLE
Kokkos_MemoryPool.hpp: include iostream

### PR DESCRIPTION
**Pull request by ndellingwood**
address build errors when profiling disabled

### CODE REVIEW

**Comment by ndellingwood** [comment]
Following merge of #3098 clang/9 and gcc/9.2 nightly builds with profiling disabled began failing with messages like below, due to removal of `iostream` from ` core/src/Kokkos_ExecPolicy.hpp`

Sample error
```bash
/workspace/Kokkos_develop_kokkosdev2_gcc920_pthreads_disable_profile/kokkos/core/src/Kokkos_MemoryPool.hpp:236:7: error: no match for ‘operator<<’ (operand types are ‘std::ostream’ {aka ‘std::basic_ostream<char>’} and ‘const char [11]’)
   236 |     s << "pool_size(" << (size_t(m_sb_count) << m_sb_size_lg2) << ")"
       |     ~ ^~ ~~~~~~~~~~~~
       |     |    |
       |     |    const char [11]
       |     std::ostream {aka std::basic_ostream<char>}
```

Sample reproducer
```bash
  #   Load modules:
        module load sems-env sems-cmake/3.12.2 sems-gcc/9.2.0

        $KOKKOS_PATH/generate_makefile.bash --with-devices=Pthread --arch=SNB,Volta70 --compiler=/projects/sems/install/rhel7-x86_64/sems/compiler/gcc/9.2.0/base/bin/g++ --cxxflags="-O3 -Wall -Wunused-parameter -Wshadow -pedantic -Werror -Wsign-compare -Wtype-limits -Wignored-qualifiers -Wempty-body -Wclobbered -Wuninitialized " --cxxstandard="11" --ldflags=""  --kokkos-path=$KOKKOS_PATH --with-options=disable_profiling --with-cuda-options= --no-examples 
```

**Comment by ndellingwood** [comment]
@dalg24 @masterleinad would it be feasible to disable profiling in one of the PR tests to make sure corresponding code paths are covered?

**Comment by DavidPoliakoff** [comment]
Good catch. We should merge this, though note that the profiling-off branch won't exist soon

File: core/src/Kokkos_MemoryPool.hpp [hunk 1]
```diff
@@ -52,6 +52,8 @@
 #include <impl/Kokkos_Error.hpp>
 #include <impl/Kokkos_SharedAlloc.hpp>
 
+#include <iostream>
```
Comment (dalg24) [inline core/src/Kokkos_MemoryPool.hpp:55]:
```suggestion
#include <iosfwd>
```

Comment (ndellingwood) [reply core/src/Kokkos_MemoryPool.hpp]: Force-pushed the change.

**Comment by dalg24** [comment]
> @dalg24 @masterleinad would it be feasible to disable profiling in one of the PR tests to make sure corresponding code paths are covered?

Sure you can change in https://github.com/kokkos/kokkos/blob/6d6b480ab68e883ddb265c8e7b91b0b02b70d7b5/.jenkins

**Comment by ndellingwood** [comment]
@dalg24 looks like profiling is already disabled in the `CUDA-9.2-NVCC` build, this issue only slipped through in host builds. May I modify the `GCC-4.8.4` jenkins test to disable profiling as part of this PR or would you prefer a separate PR to modify the jenkins file?

**Comment by dhollman** [comment]
We have to be a little careful about just replacing `<iostream>` with `<iosfwd>` when inline function templates and class templates are defined like this. It will break code for some people (who, in fairness, already had broken code) that are relying on Kokkos as a means of transitively importing `<iostream>`. If we want to just say not a defect, that's fine, but we should be clear about it.

**Comment by ndellingwood** [comment]
@dhollman should this be closed in favor of #3109?

**Comment by ndellingwood** [comment]
Tester didn't like iosfwd, though the change built for me locally with gcc-4.8.4. I pushed change replacing with iostream, though this PR is likely unnecessary and to be replaced by #3109

**Comment by masterleinad** [comment]
> Tester didn't like iosfwd, though the change built for me locally with gcc-4.8.4. I pushed change replacing with iostream, though this PR is likely unnecessary and to be replaced by #3109

AFAICT, #3109 doesn't touch `Kokkos_Memeory_Pool.h`, does it?

**Comment by ndellingwood** [comment]
> AFAICT, #3109 doesn't touch `Kokkos_Memeory_Pool.h`, does it?

You're correct, I skimmed to quickly over the github notification of the submitted PR and misread the filename (it touches Kokkos_MemorySpace.hpp). Thanks for the catch.

**Comment by ndellingwood** [comment]
Looks like #3113 just went in that will make this PR irrelevant.

**Comment by ndellingwood** [comment]
cuda-9.2 failure requiring merge of PR #3114

**Comment by dalg24** [comment]
Retest this please

File: .jenkins [hunk 2]
```diff
@@ -238,6 +238,7 @@ pipeline {
                                 -DKokkos_ENABLE_COMPILER_WARNINGS=ON \
                                 -DKokkos_ENABLE_TESTS=ON \
                                 -DKokkos_ENABLE_OPENMP=ON \
+                                -DKokkos_ENABLE_PROFILING=OFF \
```
Comment (masterleinad) [inline .jenkins:241]: We need to make sure to change this in #3115.

**Pull request merged**